sealedenumprocessor/src/main/kotlin/com/livefront/sealedenum/internal/SuperInterfaces.kt

```diff
@@ -0,0 +1,124 @@
+package com.livefront.sealedenum.internal
+
+import com.squareup.kotlinpoet.ANY
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.Dynamic
+import com.squareup.kotlinpoet.LambdaTypeName
+import com.squareup.kotlinpoet.ParameterizedTypeName
+import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
+import com.squareup.kotlinpoet.STAR
+import com.squareup.kotlinpoet.TypeName
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.TypeVariableName
+import com.squareup.kotlinpoet.WildcardTypeName
+import com.squareup.kotlinpoet.metadata.KotlinPoetMetadataPreview
+import com.squareup.kotlinpoet.metadata.specs.ClassInspector
+import com.squareup.kotlinpoet.metadata.specs.classFor
+import com.squareup.kotlinpoet.metadata.specs.toTypeSpec
+
+/**
+ * Given a [typeSpec] and the type names for its type parameters, returns a list of all valid super interfaces that it
+ * or any of its superclasses implement.
+ *
+ * A super interface is "valid" if it does not require a wildcard generic type parameter.
+ */
+@KotlinPoetMetadataPreview
+internal fun ClassInspector.getAllSuperInterfaces(typeSpec: TypeSpec): List<TypeName> {
+    val superInterfaces = mutableListOf<TypeName>()
+    getAllSuperInterfacesImpl(
+        typeSpec,
+        typeSpec.wildcardedTypeVariables,
+        superInterfaces
+    )
+    return superInterfaces
+}
+
+/**
+ * The recursive implementation for [getAllSuperInterfaces].
+ *
+ * This method adds all valid super interfaces to [superInterfaces].
+ */
+@KotlinPoetMetadataPreview
+private tailrec fun ClassInspector.getAllSuperInterfacesImpl(
```

Comment:
Ooo tail recursion....

Reply:
It was only 50% because it seemed cool!

I originally started to make the iterative version, but it seemed a lot clearer to let it be handled recursively.

Reply:
I agree. Recursion for the win.